File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/SimplyJson.java
package com.wwdablu.soumya.simplypdf.jsonengine;

import android.content.Context;

import androidx.annotation.NonNull;

import com.wwdablu.soumya.simplypdf.SimplyPdfDocument;
import com.wwdablu.soumya.simplypdf.composers.ImageComposer;
import com.wwdablu.soumya.simplypdf.composers.ShapeComposer;
import com.wwdablu.soumya.simplypdf.composers.TableComposer;
import com.wwdablu.soumya.simplypdf.composers.TextComposer;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;

public final class SimplyJson {

    private Context context;
    private String payload;
    private SimplyPdfDocument simplyPdfDocument;

    private ObservableEmitter<Boolean> emitter;

    public SimplyJson(@NonNull Context context, @NonNull String payload) {
        this.context = context;
        this.payload = payload;
    }

    public Observable<Boolean> generateWith(@NonNull SimplyPdfDocument simplyPdfDocument) {

        this.simplyPdfDocument = simplyPdfDocument;
        return Observable.create(e -> {
            emitter = e;

            try {
                parsePayload();
            } catch (Exception ex) {
                emitter.onError(ex);
            }
        });
    }

    private void emitSuccess() throws IOException {
        simplyPdfDocument.finish();
        emitter.onNext(true);
        emitter.onComplete();
    }

    private void emitFailure(Throwable t) {
        emitter.onError(t);
    }

    private void parsePayload() throws Exception {

        JSONObject root = new JSONObject(payload);
        JSONArray contentArray = root.getJSONArray(Node.CONTENT);

        if(contentArray == null || contentArray.length() == 0) {
            emitSuccess();
            return;
        }

        TextComposer textComposer = null;
        ImageComposer imageComposer = null;
        TableComposer tableComposer = null;
        ShapeComposer shapeComposer = null;

        TextConverter textConverter = null;
        ImageConverter imageConverter = null;
        TableConverter tableConverter = null;
        ShapeConverter shapeConverter = null;

        for(int index = 0; index < contentArray.length(); index++) {

            JSONObject compose = contentArray.getJSONObject(index);
            if(compose == null) {
                emitFailure(new IllegalArgumentException("Could not parse JSON"));
                return;
            }

            switch (compose.getString(Node.TYPE).toLowerCase()) {

                case Node.TYPE_TEXT:
                    if(textComposer == null) {
                        textComposer = new TextComposer(simplyPdfDocument);
                        textConverter = new TextConverter();
                    }

                    textConverter.generate(textComposer, compose);
                    break;

                case Node.TYPE_IMAGE:
                    if(imageComposer == null) {
                        imageComposer = new ImageComposer(simplyPdfDocument);
                        imageConverter = new ImageConverter(context);
                    }

                    imageConverter.generate(imageComposer, compose);
                    break;

                case Node.TYPE_TABLE:
                    if(tableComposer == null) {
                        tableComposer = new TableComposer(simplyPdfDocument);
                        tableConverter = new TableConverter();
                    }

                    tableConverter.generate(tableComposer, compose);
                    break;

                case Node.TYPE_SHAPE:
                    if(shapeComposer == null) {
                        shapeComposer = new ShapeComposer(simplyPdfDocument);
                        shapeConverter = new ShapeConverter();
                    }

                    shapeConverter.generate(shapeComposer, compose);
                    break;

                case Node.TYPE_SPACE:
                    simplyPdfDocument.insertEmptySpace(compose.getInt(Node.COMPOSER_SPACE_HEIGHT));
                    break;

                case Node.TYPE_NEWPAGE:
                    simplyPdfDocument.newPage();
                    break;
            }
        }

        emitSuccess();
    }
}

File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/SimplyJson.kt
package com.wwdablu.soumya.simplypdf.jsonengine

import android.content.Context
import com.wwdablu.soumya.simplypdf.SimplyPdf
import com.wwdablu.soumya.simplypdf.SimplyPdfDocument
import com.wwdablu.soumya.simplypdf.jsonengine.base.ComposerConverter
import com.wwdablu.soumya.simplypdf.jsonengine.base.PageConverter
import com.wwdablu.soumya.simplypdf.jsonengine.composerconverters.*
import com.wwdablu.soumya.simplypdf.jsonengine.pagemodifiers.HeaderConverter
import com.wwdablu.soumya.simplypdf.jsonengine.setup.SetupHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

internal class SimplyJson(private val context: Context, private val payload: String) {
    
    private lateinit var simplyPdfDocument: SimplyPdfDocument

    private val pageConverters: HashMap<String, PageConverter> by lazy { HashMap<String, PageConverter>().also { map ->
            HeaderConverter(simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
        }
    }

    private val composerConverters: HashMap<String, ComposerConverter> by lazy { HashMap<String, ComposerConverter>().also { map ->
            TextConverter(simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
            ImageConverter(context, simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
            TableConverter(simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
            NewPageConverter(simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
            SpaceConverter(simplyPdfDocument).apply {
                map[getTypeHandler()] = this
            }
        }
    }

    suspend fun generateWith(pdfBuilder: SimplyPdf,
                             pageConverters: List<PageConverter>? = null,
                             composerConverters: List<ComposerConverter>? = null) {

        val contentArray: JSONArray? = JSONObject(payload).getJSONArray(Node.PAGE)
        if (contentArray == null || contentArray.length() == 0) {
            return
        }

        for(index in 0 until contentArray.length()) {
            if((contentArray.getJSONObject(index).has(Node.TYPE)) &&
                (contentArray.getJSONObject(index).getString(Node.TYPE).lowercase() == Node.TYPE_PAGE_SETUP)) {

                generateWith(SetupHandler(pdfBuilder, contentArray.getJSONObject(index).toString())
                    .apply(), pageConverters, composerConverters)
                return
            }
        }

        generateWith(pdfBuilder.build(), pageConverters, composerConverters)
    }

    @JvmOverloads
    suspend fun generateWith(simplyPdfDocument: SimplyPdfDocument,
                             pageConverters: List<PageConverter>? = null,
                             composerConverters: List<ComposerConverter>? = null) {

        this.simplyPdfDocument = simplyPdfDocument

        pageConverters?.forEach {
            this.pageConverters[it.getTypeHandler()] = it
        }

        composerConverters?.forEach {
            this.composerConverters[it.getTypeHandler()] = it
        }

        withContext(Dispatchers.IO) {
            parsePagePayload()
            parseContentPayload()
            simplyPdfDocument.finish()
        }
    }

    private fun parsePagePayload() {

        val contentArray: JSONArray? = JSONObject(payload).getJSONArray(Node.PAGE)
        if (contentArray == null || contentArray.length() == 0) {
            return
        }

        for (index in 0 until contentArray.length()) {
            val compose: JSONObject = contentArray.getJSONObject(index) ?: continue
            pageConverters[compose.getString(Node.TYPE).lowercase()]?.generate(compose)
        }
    }

    private fun parseContentPayload() {

        val contentArray: JSONArray? = JSONObject(payload).getJSONArray(Node.CONTENTS)
        if (contentArray == null || contentArray.length() == 0) {
            return
        }

        for (index in 0 until contentArray.length()) {
            val compose: JSONObject = contentArray.getJSONObject(index) ?: continue
            composerConverters[compose.getString(Node.TYPE).lowercase()]?.generate(compose)
        }
    }
}